src/piton/labelers/core.py

```diff
@@ -0,0 +1,381 @@
+        self.time = time
+        self.label_type = label_type
+        self.value = value
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Label):
+            return NotImplemented
+        return (
+            self.time,
+            self.label_type,
+            self.value,
+        ) == (
+            other.time,
+            other.label_type,
+            other.value,
+        )
+
+    def __repr__(self) -> str:
+        return f"Label(time={self.time}, value={self.value}, type={self.label_type})"
+
+class LabelingFunction(ABC):
+    """
+    An interface for labeling functions.
+    
+    A labeling function applies a label to a specific datetime in a given patient's timeline.
+    It can be thought of as generating the following list given a specific patient:
+        [(patient ID, datetime_1, label_1), (patient ID, datetime_2, label_2), ... ]
+    
+    Usage:
+        labeling_function: LabelingFunction = LF(...)
+        patients: Sequence[Patient] = ...
+        labels: LabeledPatient = labeling_function.apply(patients)
+    """
+
+    @abstractmethod
+    def label(self, patient: Patient) -> List[Label]:
+        """
+        Applies every label that is applicable to the provided patient.
+        This is only called once per patient.
+
+        Args:
+            patient (Patient): A patient object
+
+        Returns:
+            List[Label]: A list of :class:`Label` containing every label for the given patient
+        """
+        pass
+
+    def get_required_codes(self) -> List[int]:
+        """Set of codes that a patient must have at least one of in order to qualify for this labeler
+        This allows us to only extract patients from the :class:`PatientDatabase` who have a code that matches one of these
+
+        Returns:
+            List[int]: List of applicable OMOP codes
+        """
+        pass
+
+    def get_patient_start_end_times(self, patient: Patient) -> Tuple[datetime.datetime, datetime.datetime]:
+        """Returns the (start, end) of the patient timeline
+
+        Returns:
+            Tuple[datetime.datetime, datetime.datetime]: (start, end)
+        """
+        pass
+
+    @abstractmethod
+    def get_labeler_type(self) -> LabelType:
+        """Return what type of labels this labeler returns. See the Label class."""
+        pass
+
+    def apply(self, patients: Sequence[Patient]) -> LabeledPatients:
+        """Applies the `label()` function one-by-one to each Patient in a sequence of Patients
+
+        Args:
+            patients (Sequence[Patient]): A sequence of Patient objcets
+
+        Returns:
+            LabeledPatients: Maps patients to labels
+        """        
+        patients_to_labels: Dict[int, List[Label]] = {}
+        for patient in patients:
+            patients_to_labels[patient.patient_id] = self.label(patient)
+        return LabeledPatients(patients_to_labels, self.get_labeler_type())
+
+class LabeledPatients(MutableMapping):
```

Comment:
I don't think the typing is right here? Should be MutableMapping[int, Label] or something of that nature?